clash-ghc/src-ghc/Clash/GHC/ClashFlags.hs

```diff
@@ -1,327 +1,328 @@
-  -> Int
-  -> IO ()
-setInlineConstantLimit r n = modifyIORef r (\c -> c {opt_inlineConstantLimit = toEnum n})
-
-setEvaluatorFuelLimit
-  :: IORef ClashOpts
-  -> Int
-  -> IO ()
-setEvaluatorFuelLimit r n = modifyIORef r (\c -> c {opt_evaluatorFuelLimit = toEnum n})
-
-setInlineWFLimit
-  :: IORef ClashOpts
-  -> Int
-  -> IO ()
-setInlineWFLimit r n = modifyIORef r (\c -> c {opt_inlineWFCacheLimit = toEnum n})
-
-setSpecLimit :: IORef ClashOpts
-             -> Int
-             -> IO ()
-setSpecLimit r n = modifyIORef r (\c -> c {opt_specLimit = n})
-
-setDebugInvariants :: IORef ClashOpts -> IO ()
-setDebugInvariants r =
-  modifyIORef r $ \c ->
-    c { opt_debug = (opt_debug c) { dbg_invariants = True } }
-
-setDebugCountTransformations :: IORef ClashOpts -> IO ()
-setDebugCountTransformations r =
-  modifyIORef r $ \c ->
-    c { opt_debug = (opt_debug c) { dbg_countTransformations = True } }
-
-setDebugTransformations :: IORef ClashOpts -> String -> EwM IO ()
-setDebugTransformations r s =
-  liftEwM (modifyIORef r (setTransformations transformations))
  where
-  transformations = Set.fromList (filter (not . null) (map trim (splitOn "," s)))
-  trim = dropWhileEnd isSpace . dropWhile isSpace
-
-  setTransformations xs opts =
-    opts { opt_debug = (opt_debug opts) { dbg_transformations = xs } }
-
-setDebugTransformationsFrom :: IORef ClashOpts -> Int -> EwM IO ()
-setDebugTransformationsFrom r n =
-  liftEwM (modifyIORef r (setFrom (fromIntegral n)))
+  voidLens :: Lens' ClashOpts Bool
+  voidLens = lens (const True) const
+
+  lowerCaseLens :: Lens' ClashOpts Bool
+  lowerCaseLens = opt_lowerCaseBasicIds . iso' preserveCaseToBool boolToPreserveCase
+
+  packUnpackLens :: Lens' (Maybe Text) (Maybe String)
+  packUnpackLens = iso' (fmap Text.unpack) (fmap Text.pack)
+
+hdlSynFlag :: FlagName -> ClashFlag
+hdlSynFlag flagName = ClashFlag
+  { cfFlags = \ref -> strSuffix flagName ref parser (set opt_hdlSyn)
+  , cfRender = \opts ->
+      if isDefault opt_hdlSyn opts
+      then []
+      else renderSepSuffix flagName (show (opts ^. opt_hdlSyn))
+  }
  where
-  setFrom from opts =
-    opts { opt_debug = (opt_debug opts) { dbg_transformationsFrom = Just from } }
-
-setDebugTransformationsLimit :: IORef ClashOpts -> Int -> EwM IO ()
-setDebugTransformationsLimit r n =
-  liftEwM (modifyIORef r (setLimit (fromIntegral n)))
+  parser :: String -> EwM IO (Maybe HdlSyn)
+  parser s = case readMaybe s of
+    Just hdlSyn ->
+      pure (Just hdlSyn)
+
+    Nothing -> case s of
+      -- TODO: Should we add a deprecation warning for these flags? We could
+      --       just use 'maybeReadableFlag' and ditch this custom impl if we
+      --       once the warning has been up long enough.
+      "Xilinx" -> pure (Just Vivado)
+      "ISE"    -> pure (Just Vivado)
+      "Altera" -> pure (Just Quartus)
+      "Intel"  -> pure (Just Quartus)
+      _ -> do
+        addErr (s ++ " is an unknown hdl synthesis tool")
+        pure Nothing
```

Comment:
While we're here: I think we should

  * make this comparison case-insensitive, since it costs us nothing but potentially makes Clash less annoying
  * recognize the names `vivado` and `quartus` since it's weird to me we use those names internally but don't accept them on the command line

Reply:
`Vivado` and `Quartus` are recognized. It only takes this branch if the `readMaybe` fails. Agreed on the lower case stuff, but this is true for all the "readable" flags. Let's postpone until we have a way to do those case-insensitively too?